app/src/main/java/org/mozilla/fenix/settings/biometric/BiometricPromptPreferenceFragment.kt

```diff
@@ -0,0 +1,121 @@
+import android.app.KeyguardManager
+import android.content.Context
+import android.content.Intent
+import android.view.View
+import androidx.core.content.getSystemService
+import androidx.preference.Preference
+import androidx.preference.PreferenceFragmentCompat
+import mozilla.components.support.base.feature.ViewBoundFeatureWrapper
+import org.mozilla.fenix.ext.settings
+import org.mozilla.fenix.settings.requirePreference
+
+/**
+ * Helper for creating and implementing the [BiometricPromptFeature]. Currently used
+ * for logins and credit cards.
+ */
+abstract class BiometricPromptPreferenceFragment : PreferenceFragmentCompat() {
+
+    internal val biometricPromptFeature = ViewBoundFeatureWrapper<BiometricPromptFeature>()
+
+    /**
+     * Gets the string to be used for [BiometricPromptFeature.requestAuthentication] prompting to
+     * unlock the device.
+     */
+    abstract fun unlockMessage(): String
+
+    /**
+     * Navigate when authentication is successful.
+     */
+    abstract fun navigateOnSuccess()
+
+    /**
+     * Show a warning to set up a pin/password when the device is not secured. This is only used
+     * when BiometricPrompt is unavailable on the device.
+     */
+    abstract fun showPinDialogWarning(context: Context)
+
+    /**
+     * Toggle preferences to enable or disable navigation during authentication flows.
+     *
+     * @param prefList a list of [Preference]s to toggle.
+     * @param enabled whether or not the preferences should be enabled.
+     */
+    fun togglePrefsEnabled(prefList: List<Int>, enabled: Boolean) {
+        for (preference in prefList) {
+            requirePreference<Preference>(preference).isEnabled = enabled
+        }
+    }
+
+    /**
+     * Create a prompt to confirm the device's pin/password and start activity based on the result.
+     * This is only used when BiometricPrompt is unavailable on the device.
+     */
+    @Suppress("Deprecation")
+    abstract fun showPinVerification(manager: KeyguardManager)
+
+    /**
+     * Sets the biometric prompt feature.
+     *
+     * @param view the view that the prompt will be associate with.
+     * @param prefList a list of [Preference]s to toggle.
+     */
+    fun setBiometricPrompt(view: View, prefList: List<Int>) {
+        biometricPromptFeature.set(
+            feature = BiometricPromptFeature(
+                context = requireContext(),
+                fragment = this,
+                onAuthFailure = {
+                    togglePrefsEnabled(prefList, true)
+                },
+                onAuthSuccess = ::navigateOnSuccess
+            ),
+            owner = this,
+            view = view
+        )
+    }
+
+    /**
+     * Use [BiometricPromptFeature] or [KeyguardManager] to confirm device security.
+     *
+     * @param prefList a list of [Preference]s to disable while authentication is happening.
+     */
+    fun verifyCredentialsOrShowSetupWarning(context: Context, prefList: List<Int>) {
```

Comment:
I was testing this and wondering why things weren't working. We never call this when navigating to the saved cards. So, we don't actually show the biometric check. I am aware of what @grigoryk said, but I think for time being we should keep to the UX spec and re-evaluate.

Reply:
This was missed in one of the million and a half times we've gone back and forth on this PR. Apologies for that.